src/wallet/mod.rs

```diff
@@ -739,7 +740,11 @@ where
             }
             Some(_) if change_val_after_add.is_dust() => {
                 // single_recipient but the only output would be below dust limit
-                return Err(Error::InsufficientFunds); // TODO: or OutputBelowDustLimit?
+                // TODO: or OutputBelowDustLimit?
+                return Err(Error::InsufficientFunds(
+                    DUST_LIMIT_SATOSHI,
+                    change_val_after_add,
+                ));
```

Comment:
If I understand correctly, it seems like the real error here is that the user provided invalid recipients: the output amounts sum to less than the dust limit. Is that accurate?

So maybe an `InvalidRecipients` error type would be better???

Reply:
Yeah, `InsufficientFunds` is definitely misleading.

Reply:
I think specific branch is hit when `TxBuilder::set_single_recipient()` is enabled, which means that either `TxBuilder::drain_wallet()` is also enabled, or that the user specified manually which utxos to pick and then `TxBuilder::manually_selected_only()` has been set.

This error happens basically when the user asks to make a tx without change (either by spending everything or only some specific utxos) but the amount selected is not enough to even make that single output.

I agree that the error is not the best, just wanted to provide more context to discuss what this could be changed to..

Reply:
So `InsufficientFunds` is an accurate description in the `TxBuilder::drain_wallet()` case, but potentially not an accurate description in the `TxBuilder::drain_wallet()` case?

Reply:
There's also one extra complication, and that is the fact that `drain_wallet()` also takes the "unspendable" utxos into account.. so there could be cases where you do have enough funds in your wallet, but you've marked some utxos as "unspendable" and without them your balance is too low.

Also it's not technically the balance "below dust", because the fees are also taken into account. So it's basically "at this fee rate, with the utxos i'm allowed to spend, i can't make a non-dust output".

The thing about the fees is true for the manual coin selection as well, but on the other end the manual coin selection has precedence over the "unspendable" utxos, so that's ignored in that case.

Reply:
To me insufficient funds still makes sense if you've marked things as unspendable etc. It just means what was specified as spendable was insufficient so that the wallet's funds in total were insufficient. 

I don't get how it's possible to select an input, only have one output and not be able to produce a non-dust output (surely the input must also be below the dust limit then!). It does sound like `OutputBelowDustLimit` might be better for this awkward situtation.

Reply:
It's a very weird corner-case but with very high fees it could happen: basically you could consume almost entirely the inputs in fees and what's left is not enough to build one output.